File: iOS/Kalorie/Core/UseCases/UpdateFoodItemUseCase.swift
//
//  UpdateFoodItemUseCase.swift
//  Kalorie
//
//  Created by Josef Antoni on 10.09.2026.
//

import Foundation

enum UpdateFoodItemError: Error {
    case invalidCode
    case invalidName
    case invalidCalories
    case invalidWeight
    case invalidPortion(FoodPortionError)
    case changedSinceLoad

    init(_ validationError: FoodItemValidationError) {
        self = validationError.mapped(
            invalidCode: .invalidCode,
            invalidName: .invalidName,
            invalidCalories: .invalidCalories,
            invalidWeight: .invalidWeight
        ) { .invalidPortion($0) }
    }
}

protocol UpdateFoodItemUseCaseProtocol {
    func callAsFunction(_ item: FoodItemDomain, previouslyLoaded: FoodItemDomain) async throws
}

struct UpdateFoodItemUseCase: UpdateFoodItemUseCaseProtocol {

    // MARK: - Properties

    private let dataProvider: any FirestoreDataProviderProtocol
    private let authProvider: any AuthProviderProtocol

    // MARK: - Init

    init(dataProvider: any FirestoreDataProviderProtocol, authProvider: any AuthProviderProtocol) {
        self.dataProvider = dataProvider
        self.authProvider = authProvider
    }

    // MARK: - Functions

    // setAsync replaces the whole document, so a maintainer editing a stale copy (e.g. two
    // catalogue-editor sessions on the same barcode) would otherwise silently discard whatever
    // changed in between — the same class of race ADR 0029 closed for foodItemSubmissions, but
    // foodItems has no submitted_at-like field to reuse as a token, so this re-reads and compares
    // the full document instead, client-side only (no firestore.rules enforcement, same trust
    // boundary ADR 0029 already accepted for ApproveSubmissionUseCase's own re-read).
    func callAsFunction(_ item: FoodItemDomain, previouslyLoaded: FoodItemDomain) async throws {
        guard authProvider.userId != nil else { throw AuthError.notAuthenticated }
        if let validationError = FoodItemValidation.validate(item) {
            throw UpdateFoodItemError(validationError)
        }
        let current: FoodItemDTO? = try await dataProvider.loadAsync(id: item.id, from: Constants.Firestore.foodItems)
        guard current?.asDomain() == previouslyLoaded else {
            throw UpdateFoodItemError.changedSinceLoad
        }
        let dto = FoodItemDTO(item: item)
        try await dataProvider.setAsync(dto, id: item.id, in: Constants.Firestore.foodItems)
    }
}

#if DEBUG
struct UpdateFoodItemUseCaseFake: UpdateFoodItemUseCaseProtocol {

    // MARK: - Properties

    var errorToThrow: Error?

    // MARK: - Functions

    func callAsFunction(_ item: FoodItemDomain, previouslyLoaded: FoodItemDomain) async throws {
        if let errorToThrow { throw errorToThrow }
    }
}
#endif

File: Android/app/src/main/java/antoni/kalorie/core/usecases/UpdateFoodItemUseCase.kt
package antoni.kalorie.core.usecases

import antoni.kalorie.core.auth.AuthError
import antoni.kalorie.core.auth.AuthProviderProtocol
import antoni.kalorie.core.models.FoodItemDomain
import antoni.kalorie.core.models.FoodItemValidation
import antoni.kalorie.core.models.FoodItemValidationError
import antoni.kalorie.core.models.FoodPortionError
import antoni.kalorie.core.networking.FirestoreDataProviderProtocol
import antoni.kalorie.core.networking.FoodItemDTO
import antoni.kalorie.core.networking.loadAsync
import antoni.kalorie.core.networking.setAsync
import antoni.kalorie.core.utils.Constants

sealed class UpdateFoodItemError : Exception() {
    data object InvalidCode : UpdateFoodItemError()
    data object InvalidName : UpdateFoodItemError()
    data object InvalidCalories : UpdateFoodItemError()
    data object InvalidWeight : UpdateFoodItemError()
    data class InvalidPortion(val error: FoodPortionError) : UpdateFoodItemError()
    data object ChangedSinceLoad : UpdateFoodItemError()

    // MARK: - Functions

    companion object {
        fun from(validationError: FoodItemValidationError): UpdateFoodItemError = validationError.mapped(
            invalidCode = InvalidCode,
            invalidName = InvalidName,
            invalidCalories = InvalidCalories,
            invalidWeight = InvalidWeight,
        ) { InvalidPortion(it) }
    }
}

interface UpdateFoodItemUseCaseProtocol {
    suspend operator fun invoke(item: FoodItemDomain, previouslyLoaded: FoodItemDomain)
}

class UpdateFoodItemUseCase(
    private val dataProvider: FirestoreDataProviderProtocol,
    private val authProvider: AuthProviderProtocol,
) : UpdateFoodItemUseCaseProtocol {

    // MARK: - Functions

    // setAsync replaces the whole document, and foodItems has no submitted_at-like token, so the
    // full document is re-read and compared client-side to catch a concurrent maintainer edit.
    override suspend fun invoke(item: FoodItemDomain, previouslyLoaded: FoodItemDomain) {
        if (authProvider.userId == null) throw AuthError.NotAuthenticated
        FoodItemValidation.validate(item)?.let { throw UpdateFoodItemError.from(it) }
        val current: FoodItemDTO? = dataProvider.loadAsync(id = item.id, from = Constants.Firestore.FOOD_ITEMS)
        if (current?.asDomain() != previouslyLoaded) throw UpdateFoodItemError.ChangedSinceLoad
        dataProvider.setAsync(FoodItemDTO(item), id = item.id, inCollection = Constants.Firestore.FOOD_ITEMS)
    }
}
